internal/stacks/stackstate/applied_change.go

```diff
@@ -106,7 +106,12 @@ func (ac *AppliedChangeResourceInstanceObject) protosForObject(addr stackaddrs.A
 		// produce this object, using exactly the same schema.
 		return nil, nil, fmt.Errorf("cannot decode new state for %s in preparation for saving it: %w", addr, err)
 	}
-	encValue, err := plans.NewDynamicValue(obj.Value, ty)
+
+	// Remove sensitive marks from the decoded value so we can re-serialize it
+	// with MessagePack. We populate sensitive paths in the final change description
+	// by referring to objSrc.AttrSensitivePaths instead.
+	unmarkedValue, _ := obj.Value.UnmarkDeep()
```

Comment:
Things get a little quirky here due to some inconsistencies of which parts of the system are tracking sensitive values in-band vs. out-of-band, and I think when I wrote this originally I was not remembering that `objSrc.Decode` above already takes into account `objSrc.AttrSensitivePaths`, but it's nice that it does that.

To help a future reader follow the data flow here, I'd suggest using `UnmarkDeepWithPaths` here, and then using the second return value as the second argument to `terraform1.NewDynamicValue` below, instead of `objSrc.AttrSensitivePaths`.

The result should be the same because the marks on this `obj.Value` were derived from `objSrc.AttrSensitivePaths` in the first place, but I think I would find this easier to follow if we are consistent about taking both the raw value and the marks from the same place, so it's easier to see that what's going on here is correct in the sense that the encoded value and the encoded marks both came from the same `cty.Value`.

Reply:
Good call! Updated.